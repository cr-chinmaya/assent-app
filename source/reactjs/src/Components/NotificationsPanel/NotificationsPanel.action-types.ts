import { INotificationPanelListItem } from './NotificationsPanel.types';

export enum NotificationsPanelActionType {
    REQUEST_ALERTS_INFO = 'REQUEST_ALERTS_INFO',
    RECEIVE_ALERTS_INFO = 'RECEIVE_ALERTS_INFO',
    RESET_ALERTS_INFO = 'RESET_ALERTS_INFO',
    POST_ALERTS_INFO = 'POST_ALERTS_INFO',
    TOGGLE_NOTIFICATION_PANEL = 'TOGGLE_NOTIFICATION_PANEL',
    SET_UNREAD_NOTIFICATIONS = 'SET_UNREAD_NOTIFICATIONS',
}

export type NotificationsPanelAction =
    | IRequestAlertsInfo
    | IReceiveAlertsInfo
    | IResetAlertsInfo
    | IPostAlertsInfo
    | IToggleNotificationPanel
    | IClearUnreadNotifications;

export interface IRequestAlertsInfo {
    type: NotificationsPanelActionType.REQUEST_ALERTS_INFO;
}

export interface IReceiveAlertsInfo {
    type: NotificationsPanelActionType.RECEIVE_ALERTS_INFO;
    itemsUnread: INotificationPanelListItem[];
    itemsRead: INotificationPanelListItem[];
    unReadNotifications: Array<string>;
}

export interface IResetAlertsInfo {
    type: NotificationsPanelActionType.RESET_ALERTS_INFO;
}

export interface IPostAlertsInfo {
    type: NotificationsPanelActionType.POST_ALERTS_INFO;
    unReadNotifications: Array<string>;
}

export interface IToggleNotificationPanel {
    type: NotificationsPanelActionType.TOGGLE_NOTIFICATION_PANEL
}

export interface IClearUnreadNotifications {
    type: NotificationsPanelActionType.SET_UNREAD_NOTIFICATIONS
}