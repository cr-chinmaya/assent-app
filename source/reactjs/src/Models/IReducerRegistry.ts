import { ReducersMapObject, AnyAction } from 'redux';

export interface IReducerRegistry {
    getReducers(): ReducersMapObject;
    getDynamicReducers(): ReducersMapObject;

    exists(reducerName: string): boolean;

    register<S, A extends AnyAction>(
        name: string,
        reducer: (state: S, action: A) => S,
        skipIfExists?: boolean
    ): IReducerRegistry;
    registerDynamic<S, A extends AnyAction>(
        name: string,
        reducer: (state: S, action: A) => S,
        skipIfExists?: boolean,
        shouldPersist?: boolean
    ): IReducerRegistry;

    addChangeListener(name: string, fn: ReducerChangeListener): IReducerRegistry;
    removeChangeListener(name: string): IReducerRegistry;
}

export type ReducerNames = string[];

export type ReducerChangeListener = (
    reducers: ReducersMapObject,
    dynamicReducers: ReducersMapObject,
    blacklistedDynamicReducer: ReducerNames
) => void;
