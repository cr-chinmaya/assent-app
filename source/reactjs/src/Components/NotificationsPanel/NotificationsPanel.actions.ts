import {
    IRequestAlertsInfo,
    IReceiveAlertsInfo,
    IResetAlertsInfo,
    IPostAlertsInfo,
    IToggleNotificationPanel,
    IClearUnreadNotifications,
    NotificationsPanelActionType
} from './NotificationsPanel.action-types';
import { INotificationPanelListItem } from './NotificationsPanel.types';

export function requestAlertsInfo(): IRequestAlertsInfo {
    return {
        type: NotificationsPanelActionType.REQUEST_ALERTS_INFO
    };
}

export function receiveAlertsInfo(itemsUnread: INotificationPanelListItem[], itemsRead: INotificationPanelListItem[], unReadNotifications: Array<string>): IReceiveAlertsInfo {
    return {
        type: NotificationsPanelActionType.RECEIVE_ALERTS_INFO,
        itemsUnread,
        itemsRead,
        unReadNotifications
    };
}

export function resetAlertsInfo(): IResetAlertsInfo {
    return {
        type: NotificationsPanelActionType.RESET_ALERTS_INFO
    }
}

export function postAlertsInfo(unReadNotifications: Array<string>): IPostAlertsInfo {
    return {
        type: NotificationsPanelActionType.POST_ALERTS_INFO,
        unReadNotifications
    }
}

export function toggleNotificationPanel(): IToggleNotificationPanel {
    return {
        type: NotificationsPanelActionType.TOGGLE_NOTIFICATION_PANEL,

    }
}

export function clearUnreadNotifications(): IClearUnreadNotifications {
    return {
        type: NotificationsPanelActionType.SET_UNREAD_NOTIFICATIONS,
    }
}