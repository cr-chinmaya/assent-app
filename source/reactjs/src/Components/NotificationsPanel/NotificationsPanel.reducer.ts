import { INotificationsPanelState } from './NotificationsPanel.types';
import { NotificationsPanelAction, NotificationsPanelActionType } from './NotificationsPanel.action-types';

export const notificationsPanelReducerName = 'NotificationsPanelReducer';
export const notificationsPanelInitialState: INotificationsPanelState = {
    itemsUnread: [],
    itemsRead: [],
    allNotifications: [],
    unReadNotifications: [],
    unReadNotificationsCount: 0,
    isNotificationPanelOpen: false,
    isLoading: false,
    hasError: false,
};

export function notificationsPanelReducer(
    prev: INotificationsPanelState = notificationsPanelInitialState,
    action: NotificationsPanelAction
): INotificationsPanelState {
    switch (action.type) {
        case NotificationsPanelActionType.REQUEST_ALERTS_INFO:
            return {
                ...prev,
                isLoading: true,
                hasError: false,
            };
        case NotificationsPanelActionType.RECEIVE_ALERTS_INFO:
            return {
                ...prev,
                isLoading: false,
                hasError: false,
                itemsUnread: action?.itemsUnread,
                itemsRead: action?.itemsRead,
                allNotifications: action?.itemsUnread.concat(action?.itemsRead),
                unReadNotifications: action?.unReadNotifications,
                unReadNotificationsCount: action?.itemsUnread?.length,
            };
        case NotificationsPanelActionType.POST_ALERTS_INFO:
            return {
                ...prev,
                isLoading: false,
                hasError: false,
                unReadNotificationsCount: 0,
            };
        case NotificationsPanelActionType.TOGGLE_NOTIFICATION_PANEL:
            return {
                ...prev,
                isNotificationPanelOpen: !prev.isNotificationPanelOpen,
                isLoading: false,
                hasError: false,
            };
        case NotificationsPanelActionType.SET_UNREAD_NOTIFICATIONS:
            return {
                ...prev,
                hasError: false,
                isLoading: false,
                unReadNotifications: [],
            };
        case NotificationsPanelActionType.RESET_ALERTS_INFO:
            return {
                ...prev,
                isLoading: false,
                hasError: true,
                itemsUnread: [],
                itemsRead: [],
                allNotifications: [],
                unReadNotifications: [],
                unReadNotificationsCount: 0,
            };
        default:
            return prev;
    }
}
