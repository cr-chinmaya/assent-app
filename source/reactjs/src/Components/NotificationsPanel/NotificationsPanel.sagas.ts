import { SimpleEffect, getContext, put, all, call, takeLatest, select, delay } from 'redux-saga/effects';
import { IHttpClient } from '@micro-frontend-react/employee-experience/lib/IHttpClient';
import { IAuthClient } from '@micro-frontend-react/employee-experience/lib/IAuthClient';
import { ITelemetryClient } from '@micro-frontend-react/employee-experience/lib/ITelemetryClient';
import { NotificationsPanelActionType, IPostAlertsInfo } from './NotificationsPanel.action-types';
import { receiveAlertsInfo, resetAlertsInfo, toggleNotificationPanel } from './NotificationsPanel.actions';
import { setHeader } from '../Shared/Components/SagasHelper';
import { trackBusinessProcessEvent, trackException, TrackingEventId } from '../../Helpers/telemetryHelpers';
import { getStateCommonTelemetryProperties } from '../Shared/SharedComponents.selectors';
import { INotificationPanelListItem } from './NotificationsPanel.types';

function* fetchAlertsInfo(): IterableIterator<SimpleEffect<{}, {}>> {
    const telemetryClient: ITelemetryClient = yield getContext('telemetryClient');
    const authClient: IAuthClient = yield getContext('authClient');
    const stateCommonProperties = yield select(getStateCommonTelemetryProperties);
    try {
        const httpClient: IHttpClient = yield getContext('httpClient');
        const alertResponse: any = yield call([httpClient, httpClient.request], {
            url: `${__API_BASE_URL__}${__API_URL_ROOT__}/alerts`,
            resource: __RESOURCE_URL__,
            headers: setHeader(null)
        });
        let tempItemsUnread: INotificationPanelListItem[] = [];
        let tempItemsRead: INotificationPanelListItem[] = [];
        const newNotifications: Array<string> = []
        
        const getIconType = (type: string): string => {
            if(type === 'danger') return 'ReportWarning'
            if(type === 'warning') return 'IncidentTriangle'
            if(type === 'info') return 'InfoSolid'
        }

        alertResponse.data.alerts.forEach((ele: any) => {
            ele.items.map((item: any) => {
                newNotifications.push(`${item.id}`);
                if (item.isRead === false) {
                    tempItemsUnread.push({
                        itemKey: item.id,
                        displayStatus: 'new',
                        status: item.isRead ? 'read' : 'unread',
                        messageBodyText: item.message,
                        subjectIcon: getIconType(ele.severity),
                        subjectHeader: item.title,
                    })
                } else {
                    tempItemsRead.push({
                        itemKey: item.id,
                        displayStatus: 'new',
                        status: item.isRead ? 'read' : 'unread',
                        messageBodyText: item.message,
                        subjectIcon: getIconType(ele.severity),
                        subjectHeader: item.title,
                    })
                }
            })
        });
        yield put(receiveAlertsInfo(tempItemsUnread, tempItemsRead, newNotifications));
        yield delay(1000) // if not delayed , badge number is not showing
        if(tempItemsUnread?.length > 0) {
            yield put(toggleNotificationPanel());
        }
        trackBusinessProcessEvent(
            authClient,
            telemetryClient,
            'Notifications - Success',
            'Notifications.Success',
            TrackingEventId.NotificationsSuccess,
            stateCommonProperties,
        );
    } catch (error: any) {
        const exception = error.message ? new Error(error.message) : error;
        yield put(resetAlertsInfo());
        trackException(
            authClient,
            telemetryClient,
            'Notifications - Failure',
            'Notifications.Failure',
            TrackingEventId.NotificationsFailure,
            stateCommonProperties,
            exception
        );
    }
}

// have to pass array of ids of read notifications
function* updateAlertsInfo(action: IPostAlertsInfo): IterableIterator<SimpleEffect<{}, {}>> {
    const telemetryClient: ITelemetryClient = yield getContext('telemetryClient');
    const authClient: IAuthClient = yield getContext('authClient');
    const stateCommonProperties = yield select(getStateCommonTelemetryProperties);
    try {
        const httpClient: IHttpClient = yield getContext('httpClient');
        yield call([httpClient, httpClient.request], {
            url: `${__API_BASE_URL__}${__API_URL_ROOT__}/userpreference?SessionId=${telemetryClient.getCorrelationId()}`,
            method: 'POST',
            resource: __RESOURCE_URL__,
            headers: setHeader(null),
            data: { FeaturePreferenceJson: null, ReadNotificationsList: `${JSON.stringify(action.unReadNotifications)}` }
        });
        trackBusinessProcessEvent(
            authClient,
            telemetryClient,
            'Update Notifications - Success',
            'Notifications.Success',
            TrackingEventId.UpdateNotificationsSuccess,
            stateCommonProperties,
        );
    } catch (error: any) {
        const exception = error.message ? new Error(error.message) : error;
        trackException(
            authClient,
            telemetryClient,
            'Update Notifications - Failure',
            'Notifications.Failure',
            TrackingEventId.UpdateNotificationsFailure,
            stateCommonProperties,
            exception
        );
    }
}

export function* notificationsPanelSagas(): IterableIterator<{}> {
    yield all([
        takeLatest(NotificationsPanelActionType.REQUEST_ALERTS_INFO, fetchAlertsInfo),
        takeLatest(NotificationsPanelActionType.POST_ALERTS_INFO, updateAlertsInfo)
    ]);
}
