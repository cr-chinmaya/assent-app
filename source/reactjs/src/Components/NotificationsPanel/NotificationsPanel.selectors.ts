import { notificationsPanelInitialState, notificationsPanelReducerName } from './NotificationsPanel.reducer';
import { INotificationsPanelState } from './NotificationsPanel.types';

export const getIsNotficationsPanelOpen = (state: INotificationsPanelState): boolean => {
    return (
        state.dynamic?.[notificationsPanelReducerName]?.isNotificationPanelOpen ||
        notificationsPanelInitialState.isNotificationPanelOpen
    );
};
