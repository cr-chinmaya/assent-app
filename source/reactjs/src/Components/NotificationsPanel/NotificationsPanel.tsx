import * as React from 'react';
import {
    notificationsPanelReducerName,
    notificationsPanelReducer,
    notificationsPanelInitialState,
} from './NotificationsPanel.reducer';
import { INotificationsPanelState } from './NotificationsPanel.types';
import { notificationsPanelSagas } from './NotificationsPanel.sagas';
import { Reducer } from 'redux';
import { requestAlertsInfo } from './NotificationsPanel.actions';
import { useDispatch } from 'react-redux';
import { useDynamicReducer } from '@micro-frontend-react/employee-experience/lib/useDynamicReducer';
import { Context, withContext } from '@micro-frontend-react/employee-experience/lib/Context';
import { Panel } from '@fluentui/react/lib/Panel';
import { postAlertsInfo, toggleNotificationPanel, clearUnreadNotifications } from './NotificationsPanel.actions';
import { FocusZone, FocusZoneDirection } from '@fluentui/react/lib/FocusZone';
import { List } from '@fluentui/react/lib/List';
import { RenderListItem } from './RenderNotificationsListItem';
import * as notificationStyled from './NotificationsPanelStyling';
import { EmptyResults } from '../Shared/Components/EmptyResults';
import { IEmployeeExperienceContext } from '@micro-frontend-react/employee-experience/lib/IEmployeeExperienceContext';

export function NotificationPanelCustom(): React.ReactElement {
    const { useSelector, dispatch, telemetryClient } = React.useContext(
        Context as React.Context<IEmployeeExperienceContext>
    );
    const { allNotifications, unReadNotifications, isNotificationPanelOpen, itemsUnread, itemsRead } = useSelector(
        (state: INotificationsPanelState) =>
            state.dynamic?.[notificationsPanelReducerName] || notificationsPanelInitialState
    );

    return (
        <Panel
            headerText={'Notifications'}
            isOpen={isNotificationPanelOpen}
            onOpened={() => {
                if (unReadNotifications?.length > 0) {
                    dispatch(postAlertsInfo(unReadNotifications));
                }
            }}
            onDismiss={() => {
                dispatch(toggleNotificationPanel());
                dispatch(clearUnreadNotifications());
            }}
            isLightDismiss
            closeButtonAriaLabel="Close notifications panel"
        >
            {unReadNotifications?.length > 0 && itemsUnread?.length > 0 && (
                <FocusZone direction={FocusZoneDirection.vertical}>
                    <h4 className={notificationStyled.italicStyle}>{itemsUnread?.length} new alert(s)</h4>
                    <div className={notificationStyled.container} data-is-scrollable>
                        <List
                            items={itemsUnread}
                            onRenderCell={(item) => <RenderListItem item={item} status="unread" />}
                        />
                    </div>
                </FocusZone>
            )}

            {unReadNotifications?.length > 0 && itemsRead?.length > 0 && (
                <FocusZone direction={FocusZoneDirection.vertical}>
                    <h4 className={notificationStyled.italicStyle}>Older alert(s)</h4>
                    <div className={notificationStyled.container} data-is-scrollable>
                        <List items={itemsRead} onRenderCell={(item) => <RenderListItem item={item} status="read" />} />
                    </div>
                </FocusZone>
            )}

            {/* When all notifications are read */}
            {unReadNotifications?.length === 0 && allNotifications?.length > 0 && (
                <FocusZone direction={FocusZoneDirection.vertical}>
                    <h4 className={notificationStyled.italicStyle}>Older alert(s)</h4>
                    <div className={notificationStyled.container} data-is-scrollable>
                        <List
                            items={allNotifications}
                            onRenderCell={(item) => <RenderListItem item={item} status="read" />}
                        />
                    </div>
                </FocusZone>
            )}

            {/* When there are no notifications at all */}
            {allNotifications?.length === 0 && (
                <div className={notificationStyled.emptyNotifications}>
                    <EmptyResults message="No new notifications" />
                </div>
            )}
        </Panel>
    );
}
// to get notifications
export function getNotifications() {
    useDynamicReducer(
        notificationsPanelReducerName,
        notificationsPanelReducer as Reducer,
        [notificationsPanelSagas],
        false
    );

    const dispatch = useDispatch();

    React.useEffect(() => {
        dispatch(requestAlertsInfo());
    }, [dispatch]);
}
