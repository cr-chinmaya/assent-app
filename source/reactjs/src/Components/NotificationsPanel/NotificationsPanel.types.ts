import { IDefaultState } from '@micro-frontend-react/employee-experience/lib/IDefaultState';
import { notificationsPanelReducerName } from './NotificationsPanel.reducer';

export interface INotificationsPanelState extends IDefaultState {
    dynamic?: {
        [notificationsPanelReducerName]: INotificationsPanelState;
    };
}

export interface INotificationsPanelState {
    itemsUnread: INotificationPanelListItem[];
    itemsRead: INotificationPanelListItem[];
    allNotifications: INotificationPanelListItem[];
    unReadNotifications: Array<string>;
    unReadNotificationsCount: number;
    isNotificationPanelOpen: boolean;
    isLoading: boolean;
    hasError: boolean;
}

export interface INotificationPanelListItem {
    itemKey: string,
    displayStatus: string,
    status: string,
    messageBodyText: string,
    subjectIcon: string,
    subjectHeader: string,
}

export interface NProps {
    item: INotificationPanelListItem,
    status: string
}
