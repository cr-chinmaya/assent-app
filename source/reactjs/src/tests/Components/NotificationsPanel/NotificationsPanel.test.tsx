import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { NotificationPanelCustom } from '../../../Components/NotificationsPanel/NotificationsPanel';
import { IReducerRegistry } from '../../../Models/IReducerRegistry';
import { Context } from '@micro-frontend-react/employee-experience/lib/Context';
import { IEmployeeExperienceContext } from '@micro-frontend-react/employee-experience/lib/IEmployeeExperienceContext';
import { ReactReduxContext } from 'react-redux';

const reducerRegistryMock: IReducerRegistry = {
    getReducers: jest.fn(),
    getDynamicReducers: jest.fn(),
    exists: () => false,
    register: jest.fn(),
    registerDynamic: jest.fn(),
    addChangeListener: jest.fn(),
    removeChangeListener: jest.fn()
};

describe('NotificationsPanel', () => {
    it('should NotificationsPanel renders correctly', () => {
        const NotificationStore = {
            allNotifications: [
                {
                    itemKey: '1',
                    displayStatus: 'new',
                    status: 'read',
                    messageBodyText: 'This is test text',
                    subjectIcon: 'ReportWarning',
                    subjectHeader: 'Testing Notification One'
                },
                {
                    itemKey: '2',
                    displayStatus: 'new',
                    status: 'unread',
                    messageBodyText: 'This is test text',
                    subjectIcon: 'IncidentTriangle',
                    subjectHeader: 'Testing Notification Two'
                },
                {
                    itemKey: '3',
                    displayStatus: 'new',
                    status: 'unread',
                    messageBodyText: 'This is test text',
                    subjectIcon: 'InfoSolid',
                    subjectHeader: 'Testing Notification Three'
                }
            ],
            unReadNotifications: ['2', '3'],
            isNotificationPanelOpen: true,
            itemsUnread: [
                {
                    itemKey: '2',
                    displayStatus: 'new',
                    status: 'unread',
                    messageBodyText: 'This is test text',
                    subjectIcon: 'IncidentTriangle',
                    subjectHeader: 'Testing Notification Two'
                },
                {
                    itemKey: '3',
                    displayStatus: 'new',
                    status: 'unread',
                    messageBodyText: 'This is test text',
                    subjectIcon: 'InfoSolid',
                    subjectHeader: 'Testing Notification Three'
                }
            ],
            itemsRead: [
                {
                    itemKey: '1',
                    displayStatus: 'new',
                    status: 'read',
                    messageBodyText: 'This is test text',
                    subjectIcon: 'ReportWarning',
                    subjectHeader: 'Testing Notification One'
                }
            ]
        };
        const reduxContext: IReduxContext = {
            reducerRegistry: { ...reducerRegistryMock },
            runSaga: jest.fn(),
            dispatch: jest.fn(),
            useSelector: jest.fn().mockImplementation(() => NotificationStore),
            // eslint-disable-next-line @typescript-eslint/camelcase
            __redux_context__: ReactReduxContext
        };
        const wrapper = render(
            <ReduxContext.Provider value={reduxContext}>
                <NotificationPanelCustom />
            </ReduxContext.Provider>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('no unreadNotifications', () => {
        const NotificationStore: any = {
            allNotifications: [
                {
                    itemKey: '1',
                    displayStatus: 'new',
                    status: 'read',
                    messageBodyText: 'This is test text',
                    subjectIcon: 'ReportWarning',
                    subjectHeader: 'Testing Notification One'
                },
                {
                    itemKey: '2',
                    displayStatus: 'new',
                    status: 'unread',
                    messageBodyText: 'This is test text',
                    subjectIcon: 'IncidentTriangle',
                    subjectHeader: 'Testing Notification Two'
                },
                {
                    itemKey: '3',
                    displayStatus: 'new',
                    status: 'unread',
                    messageBodyText: 'This is test text',
                    subjectIcon: 'InfoSolid',
                    subjectHeader: 'Testing Notification Three'
                }
            ],
            unReadNotifications: [],
            isNotificationPanelOpen: true,
            itemsUnread: [],
            itemsRead: [
                {
                    itemKey: '1',
                    displayStatus: 'new',
                    status: 'read',
                    messageBodyText: 'This is test text',
                    subjectIcon: 'ReportWarning',
                    subjectHeader: 'Testing Notification One'
                }
            ]
        };
        const reduxContext: IReduxContext = {
            reducerRegistry: { ...reducerRegistryMock },
            runSaga: jest.fn(),
            dispatch: jest.fn(),
            useSelector: jest.fn().mockImplementation(() => NotificationStore),
            // eslint-disable-next-line @typescript-eslint/camelcase
            __redux_context__: ReactReduxContext
        };
        const { container, getByText } = render(
            <ReduxContext.Provider value={reduxContext}>
                <NotificationPanelCustom />
            </ReduxContext.Provider>
        );
        // const getAlertIcon = container.querySelector("button[aria-label='Alert Notifications']");
        // fireEvent.click(getAlertIcon);
        expect(getByText('Notifications')).toBeInTheDocument();
    });

    it('empty notifications', () => {
        const NotificationStore: any = {
            allNotifications: [],
            unReadNotifications: [],
            isNotificationPanelOpen: true,
            itemsUnread: [],
            itemsRead: []
        };
        const reduxContext: IReduxContext = {
            reducerRegistry: { ...reducerRegistryMock },
            runSaga: jest.fn(),
            dispatch: jest.fn(),
            useSelector: jest.fn().mockImplementation(() => NotificationStore),
            // eslint-disable-next-line @typescript-eslint/camelcase
            __redux_context__: ReactReduxContext
        };
        const { container, getByText } = render(
            <ReduxContext.Provider value={reduxContext}>
                <NotificationPanelCustom />
            </ReduxContext.Provider>
        );
        // const getAlertIcon = container.querySelector("button[aria-label='Alert Notifications']");
        // fireEvent.click(getAlertIcon);
        expect(getByText('Notifications')).toBeInTheDocument();
    });

    it('close notifications', () => {
      const NotificationStore: any = {
          allNotifications: [],
          unReadNotifications: [],
          isNotificationPanelOpen: false,
          itemsUnread: [],
          itemsRead: []
      };
      const reduxContext: IReduxContext = {
          reducerRegistry: { ...reducerRegistryMock },
          runSaga: jest.fn(),
          dispatch: jest.fn(),
          useSelector: jest.fn().mockImplementation(() => NotificationStore),
          // eslint-disable-next-line @typescript-eslint/camelcase
          __redux_context__: ReactReduxContext
      };
      const { container, getByText } = render(
          <ReduxContext.Provider value={reduxContext}>
              <NotificationPanelCustom />
          </ReduxContext.Provider>
      );
      expect(/'Notifications'/).not.toBeUndefined();
  });
});
