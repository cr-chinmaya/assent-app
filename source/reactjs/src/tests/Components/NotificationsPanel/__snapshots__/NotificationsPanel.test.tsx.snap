// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsPanel should NotificationsPanel renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="css-0"
  >
    <div>
      <span
        class="ms-layer"
      />
    </div>
    <div
      class="ms-Layer ms-Layer--fixed root-122"
      data-portal-element="true"
    >
      <div
        class="ms-Fabric ms-Layer-content content-124"
      >
        <div
          aria-labelledby="Panel1-headerText"
          aria-modal="true"
          class=""
          role="dialog"
          style="outline: none;"
        >
          <div
            aria-hidden="false"
            class="ms-Panel is-open ms-Panel--custom root-109"
            id="CoherPanel0"
          >
            <div
              class="ms-Overlay overlay-126"
            />
            <div
              class="ms-Panel-main main-111"
            >
              <div
                aria-hidden="true"
                data-is-visible="true"
                style="pointer-events: none; position: fixed;"
                tabindex="0"
              />
              <div
                class="ms-Panel-commands commands-112"
                data-is-visible="true"
              >
                <div
                  class="ms-Panel-navigation navigation-113"
                >
                  <div
                    class="backButtonContainer-102"
                  />
                </div>
              </div>
              <div
                class="ms-Panel-contentInner contentInner-114"
              >
                <h1
                  class="header-104"
                  id="Panel1-headerText"
                >
                  Notifications
                </h1>
                <div
                  class="ms-Panel-scrollableContent scrollableContent-117"
                  data-is-scrollable="true"
                >
                  <div
                    class="body-105"
                  >
                    <div
                      class="ms-FocusZone css-127"
                      data-focuszone-id="FocusZone2"
                    >
                      <span
                        class="css-100"
                      >
                        2
                         new alert(s)
                      </span>
                      <div
                        class="css-84"
                        data-is-scrollable="true"
                      >
                        <div
                          class="ms-List"
                          role="list"
                        >
                          <div
                            class="ms-List-surface"
                            role="presentation"
                          >
                            <div
                              class="ms-List-page"
                              role="presentation"
                            >
                              <div
                                class="ms-List-cell"
                                data-automationid="ListCell"
                                data-list-index="0"
                                role="listitem"
                              >
                                <div
                                  class="itemCellBold-86"
                                  data-is-focusable="true"
                                >
                                  <i
                                    aria-hidden="true"
                                    class="root-128"
                                    data-icon-name="IncidentTriangle"
                                  />
                                  <div
                                    class="itemContent-89"
                                  >
                                    <div
                                      class="itemNameBold-91"
                                    >
                                      Testing Notification Two
                                    </div>
                                    <div
                                      class="itemMessageBold-92"
                                    >
                                      <div>
                                        This is test text
                                      </div>
                                    </div>
                                  </div>
                                  <i
                                    aria-hidden="true"
                                    class="chevron-129"
                                    data-icon-name="ChevronDown"
                                  />
                                </div>
                              </div>
                              <div
                                class="ms-List-cell"
                                data-automationid="ListCell"
                                data-list-index="1"
                                role="listitem"
                              >
                                <div
                                  class="itemCellBold-86"
                                  data-is-focusable="true"
                                >
                                  <i
                                    aria-hidden="true"
                                    class="root-130"
                                    data-icon-name="InfoSolid"
                                  />
                                  <div
                                    class="itemContent-89"
                                  >
                                    <div
                                      class="itemNameBold-91"
                                    >
                                      Testing Notification Three
                                    </div>
                                    <div
                                      class="itemMessageBold-92"
                                    >
                                      <div>
                                        This is test text
                                      </div>
                                    </div>
                                  </div>
                                  <i
                                    aria-hidden="true"
                                    class="chevron-129"
                                    data-icon-name="ChevronDown"
                                  />
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ms-FocusZone css-127"
                      data-focuszone-id="FocusZone3"
                    >
                      <span
                        class="css-100"
                      >
                        Older alert(s)
                      </span>
                      <div
                        class="css-84"
                        data-is-scrollable="true"
                      >
                        <div
                          class="ms-List"
                          role="list"
                        >
                          <div
                            class="ms-List-surface"
                            role="presentation"
                          >
                            <div
                              class="ms-List-page"
                              role="presentation"
                            >
                              <div
                                class="ms-List-cell"
                                data-automationid="ListCell"
                                data-list-index="0"
                                role="listitem"
                              >
                                <div
                                  class="itemCell-87"
                                  data-is-focusable="true"
                                >
                                  <i
                                    aria-hidden="true"
                                    class="root-131"
                                    data-icon-name="ReportWarning"
                                  />
                                  <div
                                    class="itemContent-89"
                                  >
                                    <div
                                      class="itemName-90"
                                    >
                                      Testing Notification One
                                    </div>
                                  </div>
                                  <i
                                    aria-hidden="true"
                                    class="chevron-129"
                                    data-icon-name="ChevronRight"
                                  />
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="true"
                data-is-visible="true"
                style="pointer-events: none; position: fixed;"
                tabindex="0"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <span
      class="ms-layer"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
